Fix ClientOptions types

grissius opened the pull request:
Remove index signature from ChannelOptions to fix intersection error
described in #1558 which causes issues on using ClientOptions direct
fields with TypeScript.

Removing of index signature required few minor changes:
 - adding few constant types that were used throughout the app
 - using `as const` assertion in xds-client
 - using not-so-great type cast in channelOptionsEqual

Alternative solution would be removing the index signature from
ChannelOptions explicitly in ClientOptions definition, which is not
trivial and probably calls for a generic type helper.

See: https://github.com/grpc/grpc-node/issues/1558
Fixes: 1558

murgatroid99 commented:
I think the correct solution here is actually to just make the index signature have the value type `any`. It's important to allow unspecified keys to break equality because that is how subchannel sharing is determined.

grissius commented:
@murgatroid99 Thank you for feedback, I am not too familiar with the codebase. I added the _index any_ signature and removed the type obscurities derived from the previous solution.

I justk kept the addition of previously added keys apart from that, which are used throughout the codebase.

murgatroid99 reviewed (approved):
Thank you for your contribution.
